scarb/src/ops/workspace.rs

```diff
@@ -29,23 +34,140 @@ pub fn read_workspace_with_source_id<'c>(
+        Some(package)
+    } else {
+        validate_virtual_manifest(manifest_path, &toml_manifest)?;
+        None
+    };
 
-    let package = Package::new(manifest.summary.package_id, manifest_path.into(), manifest);
+    if let Some(workspace) = toml_workspace {
+        let workspace_root = manifest_path
+            .parent()
+            .expect("Manifest path must have parent.");
+
+        // Read workspace members.
+        let mut packages = workspace
+            .members
+            .map(|m| find_member_paths(workspace_root, m))
+            .unwrap_or_else(|| Ok(Vec::new()))?
+            .iter()
+            .map(AsRef::as_ref)
+            .map(|package_path| {
+                let package_manifest = TomlManifest::read_from_path(package_path)?;
+                // Read the member package.
+                let manifest = package_manifest
+                    .to_manifest(
+                        package_path,
+                        source_id,
+                        config.profile(),
+                        Some(&toml_manifest),
+                    )
+                    .with_context(|| format!("failed to parse manifest at `{manifest_path}`"))?;
+                let manifest = Box::new(manifest);
+                let package =
+                    Package::new(manifest.summary.package_id, package_path.into(), manifest);
+                Ok(package)
+            })
+            .collect::<Result<Vec<_>>>()?;
+        // Read root package.
+        let root_package = root_package.map(|p| {
+            packages.push(p.clone());
+            p.id
+        });
+        Workspace::new(
+            manifest_path.into(),
+            packages.as_ref(),
+            root_package,
+            config,
+            profiles,
+        )
+    } else {
+        // Read single package workspace
+        let package = root_package.ok_or_else(|| anyhow!("the [package] section is missing"))?;
+        Workspace::from_single_package(package, config, profiles)
+    }
+}
 
-    Workspace::from_single_package(package, config)
+fn find_member_paths(root: &Utf8Path, globs: Vec<String>) -> Result<Vec<Utf8PathBuf>> {
+    globs
+        .iter()
+        .map(|path| {
+            // Expand globs from workspace root.
+            glob(root.join(path).to_string().as_str())
+                .with_context(|| format!("could not parse pattern `{}`", &path))?
+                .map(|p| p.with_context(|| format!("unable to match path to pattern `{}`", &path)))
+                .map(|p| {
+                    // Return manifest path.
+                    p.map(|p| p.join(MANIFEST_FILE_NAME))
+                        .map(PathBuf::try_into_utf8)
+                })
+                .collect::<Result<Result<Vec<_>, _>>>()?
+        })
+        .collect::<Result<Vec<_>>>()
+        .map(|v| {
+            v.into_iter()
+                .flatten()
+                // Make sure all files exist.
+                .filter(|p| p.is_file())
```

Comment:
I wonder if we shouldn't print a warning if we matched a directory that is missing Scarb.toml

Technically, I think we should make users avoid overusing globs in this way? What do you think?

Reply:
Hmm, do you think a warning is enough, or we should just error?

Reply:
Afaik Cargo does emit errors for such cases, but they also have `workspace.exclude` for better control over the search.

Reply:
warning should be enough I think

workspace.exclude is something we can leave in our backlog :)

Reply:
this change can be done in separate PR, resolving this discussion